src/util/cilfacade.ml

```diff
@@ -410,3 +380,89 @@ let countLoc fn =
   let res = Hashtbl.length locs in
   Hashtbl.clear locs;
   res
+
+
+
+module StmtH = Hashtbl.Make (CilType.Stmt)
+
+let stmt_fundecs: fundec StmtH.t Lazy.t =
+  lazy (
+    let h = StmtH.create 113 in
+    iterGlobals !current_file (function
+        | GFun (fd, _) ->
+          List.iter (fun stmt ->
+              StmtH.replace h stmt fd
+            ) fd.sallstmts
+        | _ -> ()
+      );
+    h
+  )
+
+let find_stmt_fundec stmt = StmtH.find (Lazy.force stmt_fundecs) stmt (* stmt argument must be explicit, otherwise force happens immediately *)
+
+
+module VarinfoH = Hashtbl.Make (CilType.Varinfo)
+
+let varinfo_fundecs: fundec VarinfoH.t Lazy.t =
+  lazy (
+    let h = VarinfoH.create 111 in
+    iterGlobals !current_file (function
+        | GFun (fd, _) ->
+          VarinfoH.replace h fd.svar fd
+        | _ -> ()
+      );
+    h
+  )
+
+let find_varinfo_fundec vi = VarinfoH.find (Lazy.force varinfo_fundecs) vi (* vi argument must be explicit, otherwise force happens immediately *)
+
+
+module StringH = Hashtbl.Make (Printable.Strings)
+
+let name_fundecs: fundec StringH.t Lazy.t =
+  lazy (
+    let h = StringH.create 111 in
+    iterGlobals !current_file (function
+        | GFun (fd, _) ->
+          StringH.replace h fd.svar.vname fd
+        | _ -> ()
+      );
+    h
+  )
+
+let find_name_fundec name = StringH.find (Lazy.force name_fundecs) name (* name argument must be explicit, otherwise force happens immediately *)
```

Comment:
Maybe `find_fundec_by_name` and `find_fundec_by_varinfo` make it clearer that these things return a `fundec`?

Reply:
There's also `find_scope_fundec` with the type `varinfo -> fundec option` right now, but that finds the surrounding `fundec` for a local variable's `varinfo`, while `find_varinfo_fundec` finds the `fundec` which has the `varinfo` as its own name. This is the especially confusing pair of functions since the former could also be called `find_fundec_by_varinfo` or whatnot to indicate that the scope is a `fundec`.

I thought about being even more verbose like `find_function_varinfo_fundec` and `find_local_varinfo_fundec` etc, but it gets so extremely verbose to use. At least compared to `getFun` or `getdec` that we had before, so I wasn't sure how far to go with it.

Reply:
> There's also `find_scope_fundec` with the type `varinfo -> fundec option`

Maybe just `find_scope` would suffice here, since in CIL there is only function scope and globals?
I guess this would somewhat resolve the confusion.

Reply:
I added documentation comments to all those functions, that should clarify the differences.